mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,279 @@
+
+        for (size_t i = 0; i < RetValues.size(); i++) {
+            RetValues[i].replaceAllUsesExcept(executeRegionOp->getResults()[i], exceptions);
+        }
+
+        // 4. Outline the region
+        func::CallOp call;
+        FailureOr<func::FuncOp> outlined = outlineSingleBlockRegion(
+            builder, OriginalMultitapeFunc->getLoc(), executeRegionOp.getRegion(),
+            OriginalMultitapeFunc.getSymName().str() + "_tape_" + std::to_string(tape_number),
+            &call);
+        OutlinedFuncs.push_back(outlined);
+
+        if (failed(outlined)) {
+            return failure();
+        }
+        return success();
+    }
+
+    void runOnOperation() override
+    {
+        Operation *module = getOperation();
+        mlir::IRRewriter builder(module->getContext());
+
+        // 1. Identify the functions with multiple tapes
+        // Walk through each function and count the number of devices.
+        // In frontend when tracing (jax_tracers.py/trace_quantum_function),
+        // each tape has its own quantum.device operation attached to it
+        SmallVector<func::FuncOp> MultitapePrograms;
+        module->walk([&](func::FuncOp func) {
+            if (countTapes(func) >= 2) {
+                MultitapePrograms.push_back(func);
+            }
+        });
+
+        // Do nothing and exit for classical and single-tape programs
+        if (MultitapePrograms.empty()) {
+            return;
+        }
+
+        // 2. Parse the function into tapes
+        // total number of operation lists is number of tapes +2
+        // for classical pre and post processing
+
+        // for (auto func : MultitapePrograms){
+        func::FuncOp func =
+            MultitapePrograms[0]; // temporary! TODO: process all multitape functions
+        SmallVector<std::shared_ptr<SmallVector<Operation *>>> OpsEachTape(countTapes(func) + 2,
+                                                                           nullptr);
+
+        CollectOperationsForEachTape(func, OpsEachTape);
+        //}
+
+        // 3. Get the SSA values needed by the post processing (PP)
+        // These need to be returned by the tapes.
+        // Note that later tapes can also need values from earilier tapes.
+        SmallVector<Value> NecessaryValuesFromEarlierTapes;
+
+        // Go through all the operands of all the PP ops
+        // Find the ones that are not produced in PP itself
+        SmallVector<Operation *> PPOps = *(OpsEachTape.back());
+
+        for (Operation *op : PPOps) {
+            for (auto operand : op->getOperands()) {
+                Operation *OperandSource = operand.getDefiningOp();
+                if (std::find(PPOps.begin(), PPOps.end(), OperandSource) == PPOps.end()) {
+                    // A PP operand not produced in PP itself, must be from the tapes/preprocessing!
+                    // Need to be replaced by the tape functions' return values
+                    NecessaryValuesFromEarlierTapes.push_back(operand);
```

Comment:
You can also use here BackwardSlice to find all operations needed for this particular operand / operation. If you already have all operations in sets that will be outlined then you can just check set membership to find in which tape they are?